pipeline_dp/pld_budget_accounting.py

```diff
@@ -0,0 +1,143 @@
+        if total_epsilon is None and noise is None:
+            raise ValueError("If epsilon is not given, then noise is required.")
+        if total_epsilon is not None and total_epsilon <= 0:
+            raise ValueError(f"Epsilon must be positive, not {total_epsilon}.")
+        if noise is not None and noise > self._max_noise:
+            raise ValueError(
+                f"Noise = {noise} is larger than the maximum = {self._max_noise}."
+            )
+
+        self._total_delta = total_delta
+        self._total_epsilon = total_epsilon
+        self._noise = noise
+
+    def request_budget(self,
+                       noise_kind: NoiseKind,
+                       sensitivity: float = 1,
+                       weight: float = 1) -> MechanismSpec:
+        """Requests a budget and constructs a mechanism based on the specs.
+
+        Args:
+            noise_kind: The kind of distribution for the mechanism.
+            sensitivity: Optional; The sensitivity for the mechanism.
+            weight: Optional; The weight for the mechanism.
+
+        Returns:
+            A mechanism spec based on the parameters.
+        """
+        mechanism_spec = MechanismSpec(noise_kind, sensitivity, weight)
+        self._mechanisms.append(mechanism_spec)
+        return mechanism_spec
+
+    def compute_budgets(self):
+        """Computes the budget based on the mechanisms.
+
+        Retrieves epsilon for delta, if epsilon not given.
+        Adjusts the noise value, if epsilon is given.
+        """
+        if self._total_epsilon is None:
+            composed_pld = self.compose_distributions(self._noise)
+            epsilon = composed_pld.get_epsilon_for_delta(self._total_delta)
+            self._total_epsilon = epsilon
+        else:
+            adjusted_noise = self.compute_and_adjust()
+            self._noise = adjusted_noise
+
+    def compute_and_adjust(self) -> float:
+        """ Adjust the noise value to minimum needed.
+
+        Repeat the composition until the noise value gives a
+        new epsilon close to the given epsilon (within a threshold).
+        By increasing the noise we can decrease the epsilon.
+
+        Returns:
+            The noise value adjusted for the given epsilon.
+        """
+        threshold = 1e-4
+        adjusted_noise = self._max_noise / 2
+        pld_laplace_gaussian = self.compose_distributions(adjusted_noise)
+        pld_epsilon = pld_laplace_gaussian.get_epsilon_for_delta(
+            self._total_delta)
+        low, high = 0, self._max_noise
+        while low <= high:
+            if abs(pld_epsilon - self._total_epsilon) <= threshold:
+                return high
+            elif pld_epsilon < self._total_epsilon:
+                high = adjusted_noise
+            elif pld_epsilon > self._total_epsilon:
+                low = adjusted_noise
+            adjusted_noise = ((high - low) / 2) + low
+            pld_laplace_gaussian = self.compose_distributions(adjusted_noise)
```

Comment:
Nit: move compoting of epsilon to the top of the loop, then no need to do computation of the eps before the loop (so it would be slightly less code)

Reply:
Removed the extra computations after fixing the loop logic 👍  Thanks!